File: pkg/executor/join/join_table_meta.go
// Copyright 2024 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package join

import (
	"sync/atomic"
	"unsafe"

	"github.com/pingcap/tidb/pkg/parser/mysql"
	"github.com/pingcap/tidb/pkg/types"
	"github.com/pingcap/tidb/pkg/util/chunk"
	"github.com/pingcap/tidb/pkg/util/codec"
	"github.com/pingcap/tidb/pkg/util/collate"
	"github.com/pingcap/tidb/pkg/util/hack"
	"github.com/pingcap/tidb/pkg/util/serialization"
)

// joinTableMeta is the join table meta used in hash join v2
type joinTableMeta struct {
	// if the row has fixed length
	isFixedLength bool
	// the row length if the row is fixed length
	rowLength int
	// if the join keys has fixed length
	isJoinKeysFixedLength bool
	// the join keys length if it is fixed length
	joinKeysLength int
	// is the join key inlined in the row data, the join key can be inlined if and only if
	// 1. keyProb.canBeInlined returns true for all the keys
	// 2. there is no duplicate join keys
	isJoinKeysInlined bool
	// the length of null map, the null map include null bit for each column in the row and the used flag for right semi/outer join
	nullMapLength int
	// the column order in row layout, as described above, the save column order maybe different from the column order in build schema
	// for example, the build schema maybe [col1, col2, col3], and the column order in row maybe [col2, col1, col3], then this array
	// is [1, 0, 2]
	rowColumnsOrder []int
	// the column size of each column, -1 mean variable column, the order is the same as rowColumnsOrder
	columnsSize []int
	// the serialize mode for each key
	serializeModes []codec.SerializeMode
	// the first n columns in row is used for other condition, if a join has other condition, we only need to extract
	// first n columns from the RowTable to evaluate other condition
	columnCountNeededForOtherCondition int
	// total column numbers for build side chunk, this is used to construct the chunk if there is join other condition
	totalColumnNumber int
	// column index offset in null map, will be 1 when if there is usedFlag and 0 if there is no usedFlag
	colOffsetInNullMap int
	// keyMode is the key mode, it can be OneInt/FixedSerializedKey/VariableSerializedKey
	keyMode keyMode
	// offset to rowData, -1 for variable length, non-inlined key
	rowDataOffset int
	// fakeKeyByte is used as the fake key when current join need keep invalid key rows
	fakeKeyByte []byte
}

func (meta *joinTableMeta) getSerializedKeyLength(rowStart unsafe.Pointer) uint32 {
	return *(*uint32)(unsafe.Add(rowStart, sizeOfNextPtr+meta.nullMapLength))
}

func (meta *joinTableMeta) isReadNullMapThreadSafe(columnIndex int) bool {
	// Other goroutine will use `atomic.StoreUint32` to write to the first 32 bit in nullmap when it need to set usedFlag
	// so read from nullMap may meet concurrent write if meta.colOffsetInNullMap == 1 && (columnIndex + meta.colOffsetInNullMap < 32)
	mayConcurrentWrite := meta.colOffsetInNullMap == 1 && columnIndex < 31
	return !mayConcurrentWrite
}

// used in tests
func (meta *joinTableMeta) getKeyBytes(rowStart unsafe.Pointer) []byte {
	switch meta.keyMode {
	case OneInt64:
		return hack.GetBytesFromPtr(unsafe.Add(rowStart, meta.nullMapLength+sizeOfNextPtr), int(serialization.Uint64Len))
	case FixedSerializedKey:
		return hack.GetBytesFromPtr(unsafe.Add(rowStart, meta.nullMapLength+sizeOfNextPtr), meta.joinKeysLength)
	case VariableSerializedKey:
		return hack.GetBytesFromPtr(unsafe.Add(rowStart, meta.nullMapLength+sizeOfNextPtr+sizeOfElementSize), int(meta.getSerializedKeyLength(rowStart)))
	default:
		panic("unknown key match type")
	}
}

func (meta *joinTableMeta) advanceToRowData(matchedRowInfo *matchedRowInfo) {
	if meta.rowDataOffset == -1 {
		// variable length, non-inlined key
		matchedRowInfo.buildRowOffset = sizeOfNextPtr + meta.nullMapLength + sizeOfElementSize + int(meta.getSerializedKeyLength(*(*unsafe.Pointer)(unsafe.Pointer(&matchedRowInfo.buildRowStart))))
	} else {
		matchedRowInfo.buildRowOffset = meta.rowDataOffset
	}
}

func (meta *joinTableMeta) isColumnNull(rowStart unsafe.Pointer, columnIndex int) bool {
	byteIndex := (columnIndex + meta.colOffsetInNullMap) / 8
	bitIndex := (columnIndex + meta.colOffsetInNullMap) % 8
	return *(*uint8)(unsafe.Add(rowStart, sizeOfNextPtr+byteIndex))&(uint8(1)<<(7-bitIndex)) != uint8(0)
}

// for join that need to set UsedFlag during probe stage, read from nullMap is not thread safe for the first 32 bit of nullMap, atomic.LoadUint32 is used to avoid read-write conflict
func (*joinTableMeta) isColumnNullThreadSafe(rowStart unsafe.Pointer, columnIndex int) bool {
	return atomic.LoadUint32((*uint32)(unsafe.Add(rowStart, sizeOfNextPtr)))&bitMaskInUint32[columnIndex+1] != uint32(0)
}

func (*joinTableMeta) setUsedFlag(rowStart unsafe.Pointer) {
	addr := (*uint32)(unsafe.Add(rowStart, sizeOfNextPtr))
	value := atomic.LoadUint32(addr)
	value |= usedFlagMask
	atomic.StoreUint32(addr, value)
}

func (*joinTableMeta) isCurrentRowUsed(rowStart unsafe.Pointer) bool {
	return (*(*uint32)(unsafe.Add(rowStart, sizeOfNextPtr)) & usedFlagMask) == usedFlagMask
}

type keyProp struct {
	canBeInlined  bool
	keyLength     int
	isKeyInteger  bool
	isKeyUnsigned bool
}

func getKeyProp(tp *types.FieldType) *keyProp {
	switch tp.GetType() {
	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear,
		mysql.TypeDuration:
		isKeyUnsigned := mysql.HasUnsignedFlag(tp.GetFlag())
		if tp.GetType() == mysql.TypeYear {
			// year type is always unsigned
			isKeyUnsigned = true
		} else if tp.GetType() == mysql.TypeDuration {
			// duration type is always signed
			isKeyUnsigned = false
		}
		return &keyProp{canBeInlined: true, keyLength: chunk.GetFixedLen(tp), isKeyInteger: true, isKeyUnsigned: isKeyUnsigned}
	case mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
		collator := collate.GetCollator(tp.GetCollate())
		return &keyProp{canBeInlined: collate.CanUseRawMemAsKey(collator), keyLength: chunk.VarElemLen, isKeyInteger: false, isKeyUnsigned: false}
	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
		// date related type will use uint64 as serialized key
		return &keyProp{canBeInlined: false, keyLength: int(serialization.Uint64Len), isKeyInteger: true, isKeyUnsigned: true}
	case mysql.TypeFloat:
		// float will use float64 as serialized key
		return &keyProp{canBeInlined: false, keyLength: int(serialization.Float64Len), isKeyInteger: false, isKeyUnsigned: false}
	case mysql.TypeNewDecimal:
		// Although decimal is fixed length, but its key is not fixed length
		return &keyProp{canBeInlined: false, keyLength: chunk.VarElemLen, isKeyInteger: false, isKeyUnsigned: false}
	case mysql.TypeEnum:
		if mysql.HasEnumSetAsIntFlag(tp.GetFlag()) {
			// enum int type is always unsigned
			return &keyProp{canBeInlined: false, keyLength: int(serialization.Uint64Len), isKeyInteger: true, isKeyUnsigned: true}
		}
		return &keyProp{canBeInlined: false, keyLength: chunk.VarElemLen, isKeyInteger: false, isKeyUnsigned: false}
	case mysql.TypeBit:
		// bit type is always unsigned
		return &keyProp{canBeInlined: false, keyLength: int(serialization.Uint64Len), isKeyInteger: true, isKeyUnsigned: true}
	default:
		keyLength := chunk.GetFixedLen(tp)
		return &keyProp{canBeInlined: false, keyLength: keyLength, isKeyInteger: false, isKeyUnsigned: false}
	}
}

// buildKeyIndex is the build key column index based on buildSchema, should not be nil
// otherConditionColIndex is the column index that will be used in other condition, if no other condition, will be nil
// columnsNeedConvertToRow is the column index that need to be converted to row, should not be nil
// needUsedFlag is true for outer/semi join that use outer to build
func newTableMeta(buildKeyIndex []int, buildTypes, buildKeyTypes, probeKeyTypes []*types.FieldType, columnsUsedByOtherCondition []int, outputColumns []int, needUsedFlag bool) *joinTableMeta {
	meta := &joinTableMeta{}
	meta.isFixedLength = true
	meta.rowLength = 0
	meta.totalColumnNumber = len(buildTypes)

	columnsNeedToBeSaved := make(map[int]struct{}, len(buildTypes))
	updateMeta := func(index int) {
		if _, ok := columnsNeedToBeSaved[index]; !ok {
			columnsNeedToBeSaved[index] = struct{}{}
			length := chunk.GetFixedLen(buildTypes[index])
			if length == chunk.VarElemLen {
				meta.isFixedLength = false
			} else {
				meta.rowLength += length
			}
		}
	}
	if outputColumns == nil {
		// outputColumns = nil means all the column is needed
		for index := range buildTypes {
			updateMeta(index)
		}
	} else {
		for _, index := range outputColumns {
			updateMeta(index)
		}
		for _, index := range columnsUsedByOtherCondition {
			updateMeta(index)
		}
	}

	meta.isJoinKeysFixedLength = true
	meta.joinKeysLength = 0
	meta.isJoinKeysInlined = true
	keyIndexMap := make(map[int]struct{})
	meta.serializeModes = make([]codec.SerializeMode, 0, len(buildKeyIndex))
	isAllKeyInteger := true
	varLengthKeyNumber := 0
	for index, keyIndex := range buildKeyIndex {
		keyType := buildKeyTypes[index]
		prop := getKeyProp(keyType)
		if prop.keyLength != chunk.VarElemLen {
			meta.joinKeysLength += prop.keyLength
		} else {
			meta.isJoinKeysFixedLength = false
			varLengthKeyNumber++
		}
		if !prop.canBeInlined {
			meta.isJoinKeysInlined = false
		}
		if prop.isKeyInteger {
			buildUnsigned := prop.isKeyUnsigned
			probeKeyProp := getKeyProp(probeKeyTypes[index])
			if !probeKeyProp.isKeyInteger {
				panic("build key is integer but probe key is not integer, should not happens")
			}
			probeUnsigned := probeKeyProp.isKeyUnsigned
			if (buildUnsigned && !probeUnsigned) || (probeUnsigned && !buildUnsigned) {
				meta.serializeModes = append(meta.serializeModes, codec.NeedSignFlag)
				meta.isJoinKeysInlined = false
				if meta.isJoinKeysFixedLength {
					// an extra sign flag is needed in this case
					meta.joinKeysLength++
				}
			} else {
				meta.serializeModes = append(meta.serializeModes, codec.Normal)
			}
		} else {
			if !prop.isKeyInteger {
				isAllKeyInteger = false
			}
			if prop.keyLength == chunk.VarElemLen {
				// keep var column by default for var length column
				meta.serializeModes = append(meta.serializeModes, codec.KeepVarColumnLength)
			} else {
				meta.serializeModes = append(meta.serializeModes, codec.Normal)
			}
		}
		keyIndexMap[keyIndex] = struct{}{}
	}
	if !meta.isJoinKeysFixedLength {
		meta.joinKeysLength = -1
	}
	if len(buildKeyIndex) != len(keyIndexMap) {
		// has duplicated key, can not be inlined
		meta.isJoinKeysInlined = false
	}
	if !meta.isJoinKeysInlined {
		if varLengthKeyNumber == 1 {
			// if key is not inlined and there is only one var-length key, then don't need to record the var length
			for i := 0; i < len(buildKeyIndex); i++ {
				if meta.serializeModes[i] == codec.KeepVarColumnLength {
					meta.serializeModes[i] = codec.Normal
				}
			}
		}
	} else {
		for _, index := range buildKeyIndex {
			updateMeta(index)
		}
	}
	if !meta.isFixedLength {
		meta.rowLength = 0
	}
	// construct the column order
	meta.rowColumnsOrder = make([]int, 0, len(columnsNeedToBeSaved))
	meta.columnsSize = make([]int, 0, len(columnsNeedToBeSaved))
	usedColumnMap := make(map[int]struct{}, len(columnsNeedToBeSaved))

	updateColumnOrder := func(index int) {
		if _, ok := usedColumnMap[index]; !ok {
			meta.rowColumnsOrder = append(meta.rowColumnsOrder, index)
			meta.columnsSize = append(meta.columnsSize, chunk.GetFixedLen(buildTypes[index]))
			usedColumnMap[index] = struct{}{}
		}
	}
	if meta.isJoinKeysInlined {
		// if join key is inlined, the join key will be the first columns
		for _, index := range buildKeyIndex {
			updateColumnOrder(index)
		}
	}
	meta.columnCountNeededForOtherCondition = 0
	if len(columnsUsedByOtherCondition) > 0 {
		// if join has other condition, the columns used by other condition is appended to row layout after the key
		for _, index := range columnsUsedByOtherCondition {
			updateColumnOrder(index)
		}
		meta.columnCountNeededForOtherCondition = len(usedColumnMap)
	}
	if outputColumns == nil {
		// outputColumns = nil means all the column is needed
		for index := range buildTypes {
			updateColumnOrder(index)
		}
	} else {
		for _, index := range outputColumns {
			updateColumnOrder(index)
		}
	}
	if isAllKeyInteger && len(buildKeyIndex) == 1 && meta.serializeModes[0] != codec.NeedSignFlag {
		meta.keyMode = OneInt64
	} else {
		if meta.isJoinKeysFixedLength {
			meta.keyMode = FixedSerializedKey
		} else {
			meta.keyMode = VariableSerializedKey
		}
	}
	if needUsedFlag {
		meta.colOffsetInNullMap = 1
		// If needUsedFlag == true, during probe stage, the usedFlag will be accessed by both read/write operator,
		// so atomic read/write is required. We want to keep this atomic operator inside the access of nullmap,
		// then the nullMapLength should be 4 bytes alignment since the smallest unit of atomic.LoadUint32 is UInt32
		meta.nullMapLength = ((len(columnsNeedToBeSaved) + 1 + 31) / 32) * 4
	} else {
		meta.colOffsetInNullMap = 0
		meta.nullMapLength = (len(columnsNeedToBeSaved) + 7) / 8
	}
	meta.rowDataOffset = -1
	if meta.isJoinKeysInlined {
		if meta.isJoinKeysFixedLength {
			meta.rowDataOffset = sizeOfNextPtr + meta.nullMapLength
		} else {
			meta.rowDataOffset = sizeOfNextPtr + meta.nullMapLength + sizeOfElementSize
		}
	} else {
		if meta.isJoinKeysFixedLength {
			meta.rowDataOffset = sizeOfNextPtr + meta.nullMapLength + meta.joinKeysLength
		}
	}
	if meta.isJoinKeysFixedLength && !meta.isJoinKeysInlined {
		meta.fakeKeyByte = make([]byte, meta.joinKeysLength)
	}
	return meta
}

File: rust/crates/tidb-exec/src/join_table_meta.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Dependency-closed hash-join row metadata from
//! `pkg/executor/join/join_table_meta.go`.
//!
//! This leaf ports the metadata decisions that sit immediately before hash
//! join row encoding: key mode, key inlining/fixed width, serializer mode,
//! saved-column order, and null-map alignment/thread-safety. The live TiDB
//! `FieldType`/collation registry, chunk fixed-length ABI, codec bytes,
//! unsafe row pointers, atomic used flags, and hash-join execution remain
//! external.

/// Abstract source field categories needed by the metadata decision table.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ColumnType {
    /// Signed integer family (including tiny/int/year-like fixed integers).
    Int,
    /// Unsigned integer family.
    UnsignedInt,
    /// Duration, always signed and inlineable.
    Duration,
    /// Year, always unsigned and inlineable.
    Year,
    /// Datetime/date/timestamp serialized as an unsigned fixed key.
    DateTime,
    /// Float/double serialized as a fixed-width non-integer key.
    Float,
    /// Binary string/blob, raw-memory inlineable variable key.
    BinaryString,
    /// Non-binary string key.
    String,
    /// Decimal variable serialized key.
    Decimal,
    /// Enum with its integer representation flag.
    EnumInt,
    /// Textual enum.
    Enum,
    /// Set value.
    Set,
    /// JSON value.
    Json,
    /// Bit is unsigned but never inlineable in the source.
    Bit,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
struct KeyProperty {
    can_be_inlined: bool,
    key_length: Option<usize>,
    is_integer: bool,
    is_unsigned: bool,
}

impl ColumnType {
    fn key_property(self) -> KeyProperty {
        match self {
            Self::Int => KeyProperty {
                can_be_inlined: true,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: false,
            },
            Self::UnsignedInt => KeyProperty {
                can_be_inlined: true,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: true,
            },
            Self::Duration => KeyProperty {
                can_be_inlined: true,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: false,
            },
            Self::Year => KeyProperty {
                can_be_inlined: true,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: true,
            },
            Self::DateTime => KeyProperty {
                can_be_inlined: false,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: true,
            },
            Self::Float => KeyProperty {
                can_be_inlined: false,
                key_length: Some(8),
                is_integer: false,
                is_unsigned: false,
            },
            Self::BinaryString => KeyProperty {
                can_be_inlined: true,
                key_length: None,
                is_integer: false,
                is_unsigned: false,
            },
            Self::String | Self::Decimal | Self::Enum | Self::Set | Self::Json => KeyProperty {
                can_be_inlined: false,
                key_length: None,
                is_integer: false,
                is_unsigned: false,
            },
            Self::EnumInt => KeyProperty {
                can_be_inlined: false,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: true,
            },
            Self::Bit => KeyProperty {
                can_be_inlined: false,
                key_length: Some(8),
                is_integer: true,
                is_unsigned: true,
            },
        }
    }

    fn fixed_row_length(self) -> Option<usize> {
        match self {
            Self::String
            | Self::BinaryString
            | Self::Decimal
            | Self::Enum
            | Self::Set
            | Self::Json => None,
            _ => Some(8),
        }
    }
}

/// Serializer mode for one join key.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum SerializeMode {
    /// Normal codec encoding.
    Normal,
    /// Mixed signed/unsigned integer encoding needs a sign marker.
    NeedSignFlag,
    /// Variable values retain a length marker to disambiguate key columns.
    KeepVarColumnLength,
}

/// Physical key representation selected for a join table.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum KeyMode {
    /// One compatible integer key can be used directly.
    OneInt64,
    /// All serialized key components have fixed widths.
    FixedSerializedKey,
    /// At least one key component has variable width.
    VariableSerializedKey,
}

/// Metadata decisions used by hash-join row layout.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct JoinTableMeta {
    /// Whether all saved build columns have fixed row width.
    pub is_fixed_length: bool,
    /// Sum of fixed saved-column widths, or zero for variable rows.
    pub row_length: usize,
    /// Whether key bytes can be stored inline in the row.
    pub is_join_keys_inlined: bool,
    /// Whether serialized key bytes have fixed width.
    pub is_join_keys_fixed_length: bool,
    /// Fixed key width, or `-1` for variable keys.
    pub join_keys_length: isize,
    /// Serializer mode for each build key.
    pub serialize_modes: Vec<SerializeMode>,
    /// Saved build-column order in the row layout.
    pub row_columns_order: Vec<usize>,
    /// Number of bytes reserved for null/used flags.
    pub null_map_length: usize,
    /// Used-flag bit offset in the null map.
    pub col_offset_in_null_map: usize,
    /// Number of columns needed for an additional join condition.
    pub column_count_needed_for_other_condition: usize,
    /// Number of build-side columns.
    pub total_column_number: usize,
    /// Selected key mode.
    pub key_mode: KeyMode,
}

impl JoinTableMeta {
    /// Builds metadata from dependency-closed source field categories.
    #[must_use]
    pub fn new(
        build_key_index: &[usize],
        build_types: &[ColumnType],
        build_key_types: &[ColumnType],
        probe_key_types: &[ColumnType],
        columns_used_by_other_condition: Option<&[usize]>,
        output_columns: Option<&[usize]>,
        need_used_flag: bool,
    ) -> Self {
        let mut meta = Self {
            is_fixed_length: true,
            row_length: 0,
            is_join_keys_inlined: true,
            is_join_keys_fixed_length: true,
            join_keys_length: 0,
            serialize_modes: Vec::with_capacity(build_key_index.len()),
            row_columns_order: Vec::new(),
            null_map_length: 0,
            col_offset_in_null_map: 0,
            column_count_needed_for_other_condition: 0,
            total_column_number: build_types.len(),
            key_mode: KeyMode::VariableSerializedKey,
        };

        let mut saved = vec![false; build_types.len()];
        let mut save_column = |index: usize| {
            if !saved[index] {
                saved[index] = true;
                if let Some(length) = build_types[index].fixed_row_length() {
                    meta.row_length += length;
                } else {
                    meta.is_fixed_length = false;
                }
            }
        };
        if let Some(output) = output_columns {
            for &index in output {
                save_column(index);
            }
            if let Some(other) = columns_used_by_other_condition {
                for &index in other {
                    save_column(index);
                }
            }
        } else {
            for index in 0..build_types.len() {
                save_column(index);
            }
        }

        let mut all_integer = true;
        let mut variable_key_count = 0usize;
        let mut seen_key_indices = std::collections::BTreeSet::new();
        for (index, &key_index) in build_key_index.iter().enumerate() {
            let build_prop = build_key_types[index].key_property();
            let probe_prop = probe_key_types[index].key_property();
            if let Some(length) = build_prop.key_length {
                meta.join_keys_length += length as isize;
            } else {
                meta.is_join_keys_fixed_length = false;
                variable_key_count += 1;
            }
            if !build_prop.can_be_inlined {
                meta.is_join_keys_inlined = false;
            }
            if build_prop.is_integer {
                assert!(
                    probe_prop.is_integer,
                    "integer build key requires integer probe key"
                );
                if build_prop.is_unsigned != probe_prop.is_unsigned {
                    meta.serialize_modes.push(SerializeMode::NeedSignFlag);
                    meta.is_join_keys_inlined = false;
                    if meta.is_join_keys_fixed_length {
                        meta.join_keys_length += 1;
                    }
                } else {
                    meta.serialize_modes.push(SerializeMode::Normal);
                }
            } else {
                all_integer = false;
                if build_prop.key_length.is_none() {
                    meta.serialize_modes
                        .push(SerializeMode::KeepVarColumnLength);
                } else {
                    meta.serialize_modes.push(SerializeMode::Normal);
                }
            }
            if !seen_key_indices.insert(key_index) {
                meta.is_join_keys_inlined = false;
            }
        }
        if !meta.is_join_keys_fixed_length {
            meta.join_keys_length = -1;
        }
        if !meta.is_join_keys_inlined && variable_key_count == 1 {
            for mode in &mut meta.serialize_modes {
                if *mode == SerializeMode::KeepVarColumnLength {
                    *mode = SerializeMode::Normal;
                }
            }
        }
        meta.key_mode = if all_integer
            && build_key_index.len() == 1
            && meta.serialize_modes.first() != Some(&SerializeMode::NeedSignFlag)
        {
            KeyMode::OneInt64
        } else if meta.is_join_keys_fixed_length {
            KeyMode::FixedSerializedKey
        } else {
            KeyMode::VariableSerializedKey
        };

        if meta.is_join_keys_inlined {
            for &index in build_key_index {
                save_column(index);
            }
        }
        if !meta.is_fixed_length {
            meta.row_length = 0;
        }
        meta.row_columns_order = Vec::new();
        let mut used = vec![false; build_types.len()];
        if meta.is_join_keys_inlined {
            for &index in build_key_index {
                append_unique_column(&mut meta.row_columns_order, &mut used, index);
            }
        }
        if let Some(other) = columns_used_by_other_condition {
            for &index in other {
                append_unique_column(&mut meta.row_columns_order, &mut used, index);
            }
            meta.column_count_needed_for_other_condition = used.iter().filter(|&&v| v).count();
        }
        if let Some(output) = output_columns {
            for &index in output {
                append_unique_column(&mut meta.row_columns_order, &mut used, index);
            }
        } else {
            for index in 0..build_types.len() {
                append_unique_column(&mut meta.row_columns_order, &mut used, index);
            }
        }

        let saved_column_count = saved.iter().filter(|&&v| v).count();
        if need_used_flag {
            meta.col_offset_in_null_map = 1;
            meta.null_map_length = (saved_column_count + 1).div_ceil(32) * 4;
        } else {
            meta.null_map_length = saved_column_count.div_ceil(8);
        }
        meta
    }

    /// Reports whether a null-map byte can be read without atomic access.
    #[must_use]
    pub const fn is_read_null_map_thread_safe(&self, column_index: usize) -> bool {
        !(self.col_offset_in_null_map == 1 && column_index < 31)
    }
}

fn append_unique_column(order: &mut Vec<usize>, used: &mut [bool], index: usize) {
    if !used[index] {
        used[index] = true;
        order.push(index);
    }
}
